File: OSSolver/org/optimizationservices/ossolver/solver/StandardSolver.cs
using System;
using System.Threading;
using System.IO;
using System.Diagnostics;
using System.Collections;

using org.optimizationservices.oscommon.util;
using org.optimizationservices.oscommon.communicationinterface;
using org.optimizationservices.oscommon.representationparser;

namespace org.optimizationservices.ossolver.solver {
	
	/// <summary>
	/// The <c>StandardSolver</c> class is a standard solver engine. 
	/// @author Jun Ma	
	/// @version 1.0, 09/01/2005
	/// @since OS 1.0
	/// @copyright (c) 2005
	/// </summary>
	public class StandardSolver : org.optimizationservices.oscommon.localinterface.DefaultSolver{

		/// <summary>
		/// default constructor
		/// </summary>
		public StandardSolver(){
		}//constructor
		
		/// <summary>
		/// run the CG solver command that is specified in the batch file. 
		/// </summary>
		public override void solve(){
			string sOS = "window";
			string sJobID = osOption.getJobID();
		
			string sInstanceFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osil";
			string sOptionFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osol";
			string sResultFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osrl";
			OSrLWriter osrlWriter = new OSrLWriter();
			osrlWriter.setServiceURI(OSParameter.SERVICE_URI);
			osrlWriter.setServiceName(OSParameter.SERVICE_NAME);
			osrlWriter.setResultTime(DateTime.Now);

			//change starts
			string sSolverPath =  OSParameter.CODE_HOME + "solver/solver";
			if(sOS.IndexOf("window") >= 0) sSolverPath += ".exe";

			string sArguments = 
				"-osil " + sInstanceFile + 
				" -osol " + sOptionFile +
				" -osrl " + sResultFile; 

			string[] msCommandLine = {
				sSolverPath, 
				"-osil", sInstanceFile, 
				"-osol", sOptionFile,
			    "-osrl", sResultFile};
			
			//no need to change below
			bool bUseSimple = false;
			if(bUseSimple){//simple version - not registering process 
				ProcessUtil.launchAndWaitForFinish(sSolverPath, sArguments, null);
				base.osrl = IOUtil.readStringFromFile(sResultFile);
			}
			else{ //complex version - registering process
				try{
					StreamReader streamReader = null;
					ProcessStartInfo processStartInfo = new ProcessStartInfo();
					processStartInfo.RedirectStandardOutput = true;
					processStartInfo.FileName = sSolverPath;
					processStartInfo.Arguments = sArguments;
					processStartInfo.UseShellExecute = false;
					processStartInfo.WindowStyle = ProcessWindowStyle.Normal;

					Process process = Process.Start(processStartInfo);
					ArrayList vProcess = (ArrayList)OSServiceUtil.processsHashTable[sJobID];
					if(vProcess == null || vProcess.Count <= 0){
						vProcess = new ArrayList();
					}
					int iProcessID = process.Id;
					vProcess.Add(iProcessID);
					OSServiceUtil.processsHashTable.Add(sJobID, vProcess);
					streamReader = process.StandardOutput;
					string sProcessOutput = streamReader.ReadToEnd();
					process.WaitForExit();
				
					vProcess.Remove(iProcessID);
					if(vProcess.Count <= 0){
						OSServiceUtil.processsHashTable.Remove(sJobID);
					}
					streamReader.Close();
					base.osrl = IOUtil.readStringFromFile(sResultFile);
					return;
				}
				catch(Exception e){
					IOUtil.log(e.ToString(), null);
					osrlWriter.setGeneralStatusType("error");
					osrlWriter.setGeneralStatusDescription(e.Message);
					base.osrl = osrlWriter.writeToString();
					return;
				}
			}
		}//solve
	}//class StandardSolver
}//namespace

File: OSCommon/src/org/optimizationservices/oscommon/util/StandardSolver.java
/**
 * @(#)StandardSolver 1.0 03/14/2004
 *
 * Copyright (c) 2004
 */
package org.optimizationservices.oscommon.util;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.GregorianCalendar;
import java.util.Hashtable;
import java.util.Vector;

import org.optimizationservices.oscommon.localinterface.DefaultSolver;
import org.optimizationservices.oscommon.representationparser.OSrLWriter;
/**
*
* <P>The <code>StandardSolver</code>class is the standard solver engine. 
*
* </p>
* @author Robert Fourer, Jun Ma, Kipp Martin
* @version 1.0, 03/14/2004
* @since OS 1.0
*/
public class StandardSolver extends DefaultSolver{
	/**
	 * default constructor.
	 */
	public StandardSolver(){
	}//constructor

	/**
	 * run the CG solver command that is specified in the batch file. 
	 */
	public void solve(){
		String sOS = System.getProperty("os.name").toLowerCase();
		String sJobID = osOption.getJobID();
		String sOSSolver = "";
		Hashtable<String, String> otherOptimizationValues = osOption.getOtherOptimizationOptionValues();
		if(otherOptimizationValues != null){
			sOSSolver = otherOptimizationValues.get("os_solver;");
		}
		if(sOSSolver == null || sOSSolver.length() <= 0) sOSSolver = OSParameter.OS_SOLVER;
		
		String sInstanceFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osil";
		if(IOUtil.existsFileOrDir(sInstanceFile)){
			IOUtil.writeStringToFile(osil, sInstanceFile);
		}
		String sOptionFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osol";
		if(IOUtil.existsFileOrDir(sOptionFile)){
			IOUtil.writeStringToFile(osol, sOptionFile);
		}
		String sResultFile = OSParameter.TEMP_FILE_FOLDER+sJobID+".osrl";
		OSrLWriter osrlWriter = new OSrLWriter();
		osrlWriter.setServiceURI(OSParameter.SERVICE_URI);
		osrlWriter.setServiceName(OSParameter.SERVICE_NAME);
		osrlWriter.setResultTime(new GregorianCalendar());

		//change starts
		String sSolverPath =  OSParameter.CODE_HOME + "solver/OSSolverService";
		if(sOS.toLowerCase().indexOf("window") >= 0){
			sSolverPath += ".exe";
		}
		else if(sOS.toLowerCase().indexOf("linux") >= 0 || sOS.toLowerCase().indexOf("nix") >= 0){
			sSolverPath += "Linux";;	
		}
		else if(sOS.toLowerCase().indexOf("mac") >= 0){
			sSolverPath += "Mac";;
		}
		else{
			;
		}
		String sArguments = 
			" -osil " + sInstanceFile + 
			" -osol " + sOptionFile +
			" -osrl " + sResultFile + 
			" -solver" + sOSSolver; 
		
		String[] msCommandLine = {
				sSolverPath, 
				"-osil", sInstanceFile, 
				"-osol", sOptionFile,
				"-osrl", sResultFile,
				"-solver", sOSSolver};

		//no need to change below
		boolean bUseSimple = false;
		if(OSParameter.SERVICE_URI == null || OSParameter.SERVICE_URI.length() <= 0){
			bUseSimple = true;
		}
		else{
			bUseSimple = false;
		}
		if(bUseSimple){//simple version - not registering process 
			try {
				ProcessUtil.launchAndWaitForFinish(sSolverPath + " " + sArguments);
				super.osrl = IOUtil.readStringFromFile(sResultFile);				
			} 
			catch (Exception e) {
				IOUtil.log(IOUtil.exceptionStackToString(e), null);
				osrlWriter.setGeneralStatusType("error");
				osrlWriter.setGeneralStatusDescription(e.getMessage());
				super.osrl = osrlWriter.writeToString();
			}
		}
		else{ //complex version - registering process
			try{
				Process process = null;		
				process = Runtime.getRuntime().exec(msCommandLine);		
				Vector<Process> vProcess = (Vector<Process>)OSServiceUtil.processsHashTable.get(sJobID);
				if(vProcess == null || vProcess.size() <= 0){
					vProcess = new Vector<Process>();
				}
				vProcess.add(process);
				OSServiceUtil.processsHashTable.put(sJobID, vProcess);
				BufferedReader br = new BufferedReader (new InputStreamReader(process.getInputStream ()));
				boolean bCont = true;
				String sProcessOutput = "";
				String sTemp;
				while (bCont){
					sTemp = br.readLine(); 
					if(sTemp != null){
						sProcessOutput += sTemp;							
					}
					else{
						bCont = false;
					}
				}
				process.waitFor();
				vProcess.remove(process);
				if(vProcess.size() <= 0){
					OSServiceUtil.processsHashTable.remove(sJobID);
				}
				br.close();	
				super.osrl = IOUtil.readStringFromFile(sResultFile);
			}
			catch(Exception e){
				IOUtil.log(IOUtil.exceptionStackToString(e), null);
				osrlWriter.setGeneralStatusType("error");
				osrlWriter.setGeneralStatusDescription(e.getMessage());
				super.osrl = osrlWriter.writeToString();
			}
		}
		if(OSParameter.SERVICE_URI == null || OSParameter.SERVICE_URI.length() <= 0){
			IOUtil.deleteFile(sInstanceFile);
			IOUtil.deleteFile(sOptionFile);
			IOUtil.deleteFile(sResultFile);
		}
		return;

	}//solve
	
	/**
	 * main for test purposes.
	 *
	 * </p>
	 *
	 * @param argv command line arguments.
	 */
	public static void main(String[] argv){
	}//main

}//StandardSolver
